Simplify `Control` internal transform calculation

kleonc opened the pull request:
```
T // translation
R // rotation
S // scaling

Control._get_internal_transform() = T(pivot_offset) * R(rotation) * S(scale) * T(-pivot_offset)
```

Previously `T(pivot_offset) * R(rotation) * S(scale) * T(-pivot_offset)` was calculated by unnecessarily doing 1 affine inverse ($`T(pivot\_offset) = T^{-1}(-pivot\_offset)`$) and 2 full Transform2D multiplications.

Now `T(pivot_offset) * R(rotation) * S(scale)` is constructed directly, and `T(-pivot_offset)` is applied locally with a dedicated method (not performing full Transform2D multiplication).

Precision of the result is not improved / the same as before because even though before a full inverse were calculated and more multiplications / additions were done, it was just some superfluous multiplications by `1.0` / additions of `0.0 * something`, hence not introducing precision errors.

kleonc commented on scene/gui/control.cpp:
```diff
@@ -703,8 +693,9 @@ void Control::_update_canvas_item_transform() {
 Transform2D Control::get_transform() const {
 	ERR_READ_THREAD_GUARD_V(Transform2D());
-	Transform2D xform = _get_internal_transform();
-	xform[2] += get_position();
+	// M = T(position) * T(pivot_offset) * R(rotation) * S(scale) * T(-pivot_offset)
+	Transform2D xform(data.rotation, data.scale, 0.0f, data.pos_cache + data.pivot_offset);
```
Using `data.pos_cache` directly instead of `get_position()`, the difference is `get_position()` has additional thread guard. But there's already one in `get_transform`.

kleonc commented on scene/gui/control.cpp:
```diff
@@ -680,18 +680,8 @@ Size2 Control::get_parent_area_size() const {
-
 void Control::_update_canvas_item_transform() {
-	Transform2D xform = _get_internal_transform();
-	xform[2] += get_position();
+	Transform2D xform = get_transform();
```
This is the only potential difference in the behavior: `_get_internal_transform()` had no thread guard, both `get_position()` and `get_transform` have thread guard. Meaning in case of thread guard failing:
- Previously `xform` would end up with the result of `_get_internal_transform()` (`get_position()` would return `Vector2()` and hence no additional translation).
- Now `xform` would end up with `Transform2D()` as `get_transform()` would fail.

So in both cases `xform` would be incorrect (but differently). Should `_update_canvas_item_transform` have a thread guard on its own to abort further execution? :thinking:

YuriSizov commented:
If it's a performance improvement, could you provide some rough metric for how much it is improved?

kleonc commented:
> If it's a performance improvement, could you provide some rough metric for how much it is improved?

I was actually hesitant to mark it with `performance`, as I have not benchmarked it whatsoever. :smile: Marked so purely based on the info I've described in the PR description. :upside_down_face:

kleonc committed:
Simplify Control internal transform calculation

kleonc commented:
Rebased and changed to simplify only `Control::_get_internal_transform`. After #89502 it's used not only in `Control::get_transform` hence makes sense to not remove it. :upside_down_face: Plus now this PR is trivial.

The speedup should still be comparable to the previous version as the main inefficiency was within `Control::_get_internal_transform`.

aaronfranke reviewed (approved):
I didn't test it, but it makes logical sense (both the approach and the code). Assuming it works, this looks good to me.

akien-mga commented:
Thanks!
